realm/realm-library/src/main/cpp/io_realm_ObjectStoreUserStore.cpp

```diff
@@ -23,35 +23,47 @@
 
 using namespace realm;
 
-static const std::shared_ptr<SyncUser>& currentUserOrThrow() { //throws
+const char* ERR_MULTIPLE_LOGGED_IN_USERS = "Cannot be called if more that one valid, logged-in user exists.";
+const char* ERR_NO_LOGGED_IN_USER = "No user logged-in yet.";
+const char* ERR_COULD_NOT_ALLOCATE_MEMORY = "Could not allocate memory to return all users.";
+
+static const std::shared_ptr<SyncUser>& currentUserOrThrow(JNIEnv *env) //throws
+{
     std::vector<std::shared_ptr<SyncUser>> all_users = SyncManager::shared().all_users();
     if (all_users.size() > 1) {
-        std::runtime_error("cannot be called if more that one valid, logged-in user exists.");//TODO user appropriate exception
+        ThrowException(env, RuntimeError, ERR_MULTIPLE_LOGGED_IN_USERS);
     } else if (all_users.size() < 1) {
-        std::runtime_error("no user logged-in yet.");//TODO user appropriate exception
+        ThrowException(env, RuntimeError, ERR_NO_LOGGED_IN_USER);
     } else {
         return all_users.front();
     }
 }
 
-JNIEXPORT jstring JNICALL Java_io_realm_ObjectStoreUserStore_getCurrentUser
-        (JNIEnv *env, jclass) {
+/*
+ * Class:     io_realm_ObjectStoreUserStore
+ * Method:    getCurrentUser
+ * Signature: ()Ljava/lang/String;
+ */
+JNIEXPORT jstring JNICALL Java_io_realm_ObjectStoreUserStore_getCurrentUser (JNIEnv *env, jclass)
+{
     TR_ENTER()
-    const std::shared_ptr<SyncUser>& user = currentUserOrThrow();
-    if (user->state() == SyncUser::State::Active) {
-        return env->NewStringUTF(user->refresh_token().c_str());
-    } else {
-        return NULL;
-    }
+    try {
+        const std::shared_ptr<SyncUser> &user = currentUserOrThrow(env);
```

Comment:
Throwing a C++ exception in `currentUserOrThrow` will be caught by `CATCH_STD()`. If you only set the Java exception state (as in `ThrowException()`), you are not allowed to call any JNI functions. As you don't return anything in `currentUserOrThrow()` if you "throw" a Java exception, `user->state` might seg. fault - or `to_jstring()` might terminate the app.